pyiron_contrib/workflow/is_nodal.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING
+
+from pyiron_contrib.workflow.io import Signals, InputSignal, OutputSignal
+
+if TYPE_CHECKING:
+    from pyiron_base.jobs.job.extension.server.generic import Server
+
+    from pyiron_contrib.workflow.io import Inputs, Outputs
+
+
+class IsNodal(ABC):
+    """
+    A mixin class for classes that can represent nodes on a computation graph.
+    """
+
+    def __init__(self, label: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.label: str = label
+        self.running = False
+        self.failed = False
+        # TODO: Replace running and failed with a state object
+        self._server: Server | None = (
+            None  # Or "task_manager" or "executor" -- we'll see what's best
+        )
+        self.signals = self._build_signal_channels()
+
+    @property
+    @abstractmethod
+    def inputs(self) -> Inputs:
+        pass
+
+    @property
+    @abstractmethod
+    def outputs(self) -> Outputs:
+        pass
+
+    @abstractmethod
+    def update(self):
+        pass
+
+    @abstractmethod
+    def run(self):
+        pass
+
+    def _build_signal_channels(self) -> Signals:
+        signals = Signals()
+        signals.input.run = InputSignal("run", self, self.run)
+        signals.output.ran = OutputSignal("ran", self)
+        return signals
+
+    @property
+    def server(self) -> Server | None:
+        return self._server
+
+    @server.setter
+    def server(self, server: Server | None):
+        self._server = server
+
+    def disconnect(self):
+        self.inputs.disconnect()
+        self.outputs.disconnect()
+        self.signals.disconnect()
+
+    @property
+    def ready(self) -> bool:
+        return not (self.running or self.failed) and self.inputs.ready
+
+    @property
+    def connected(self) -> bool:
+        return self.inputs.connected or self.outputs.connected or self.signals.connected
```

Comment:
But a kid is an orphan only if they don't have parents and not if they don't have offsprings :p. But it's actually not just a play on words - a node has no utility only if the inputs are not complete, as far as I can imagine. In the current implementation, a node is `connected` also when only the outputs are connected. I don't know what this then allows me to know.

Reply:
> But a kid is an orphan only if they don't have parents and not if they don't have offsprings :p

Haha, good point! My english is bad 😝 

> a node has no utility only if the inputs are not complete, as far as I can imagine. In the current implementation, a node is `connected` also when only the outputs are connected. I don't know what this then allows me to know.

Not so. If only the inputs are connected, then the node is terminal and maybe (likely?) holds the target output of the workflow. If only the outputs are connected, then the node is a starting-node and allows users to influence the workflow by updating input values. Only nodes that are "only children" have guaranteed zero impact on the rest of the workflow.

What would be more useful would be graph analysis extracting unconnected subgraphs from each other, so this is just a _really_ simplified version of that that can only find single-node subgraphs.

Reply:
> If only the outputs are connected, then the node is a starting-node and allows users to influence the workflow by updating input values. Only nodes that are "only children" have guaranteed zero impact on the rest of the workflow.

But in that case the inputs will still be saturated (or there's no input at all), right?

Reply:
> But in that case the inputs will still be saturated (or there's no input at all), right?

I'm afraid I don't understand the question. What is "saturated"? Note that `connected` is not related to the `.value` stored by the channel, which may or may not be set, but only to the `.connections` list.

Reply:
Let's take this example

```python
def f(x: int) -> int:
    return x

def g(x: int) -> int:
    return x

n_f = Node(f, 'y')
n_g = Node(g, 'y')
n_g.inputs.x = n_f.outputs.y
print(n_f.connected)
```

Output: `True`

What allows me to understand by knowing that `n_f.connected` is True?

Reply:
> What allows me to understand by knowing that `n_f.connected` is True?

Fair. Very little. However, this might be useful:

```
def Workflow.get_isolated_nodes(self) -> list[Node]:
    return [node for node in self.nodes if not node.connected]
```

I'm not dead-set on having `Node.connected: bool`, but I do think it's a sensibly defined concept, and has downstream applications that are at least kind of useful. I could see getting rid of it until we have two places that use `node.input.connected and/or node.output.connected and/or node.signals.connected` -- as soon as that shows up twice I want `node.connected`. Since it currently shows up _zero_ times, if you want to get rid of it we can at least do so for now!

Reply:
But I don't really see why it's useful to know whether a node is isolated or not.

Actually it is also closely related to the fact that there's (at least as far as I know) nothing that allows me to see whether a node is ready, i.e. in the example above, after setting something like `n_f.inputs.x = 1`, I would like to see something changes, in order to know that `n_f` is ready to run.

Reply:
> But I don't really see why it's useful to know whether a node is isolated or not.

Yeah, I mean it's really only useful if you've made a big graph and you're not sure if you've left any garbage lying around in the process. So, like, I don't think it's _useless_, but I agree it's not _particularly_ useful!

> Actually it is also closely related to the fact that there's (at least as far as I know) nothing that allows me to see whether a node is ready, i.e. in the example above, after setting something like `n_f.inputs.x = 1`, I would like to see something changes, in order to know that `n_f` is ready to run.

[Node.ready](https://github.com/pyiron/pyiron_contrib/blob/5064692b9bed95a313f9615a8ce748721a39ca34/pyiron_contrib/workflow/node.py#L550)

Reply:
> Yeah, I mean it's really only useful if you've made a big graph and you're not sure if you've left any garbage lying around in the process. So, like, I don't think it's useless, but I agree it's not particularly useful!

But an isolated node is just as much of garbage as a node that has only its outputs connected, no?

> [Node.ready](https://github.com/pyiron/pyiron_contrib/blob/5064692b9bed95a313f9615a8ce748721a39ca34/pyiron_contrib/workflow/node.py#L550)

Ahaaa, now I see another problem, because if I define `f(x)` without type hinting then it shows ready even if the input is not specified. We kind of talked about this before, but I still think `Node` should not start setting an input (i.e. `None`) when the same thing shouldn't work in the first place when executed without `Node`. But as I said, it's a different problem, so maybe I'm gonna open another PR.

And while playing around, I also realized that setting `n_f.inputs.x = 1` keeps `n_f.fully_connected` false!? Eyeyey the follow-up to this will be in a different PR.